### PR TITLE
Cached tree

### PR DESCRIPTION
**Pull request by DavidVorick**
Did some housekeeping, then added functionality to support cached merkle trees.

I need to add more documentation on their usage, and probably for the package in general.

Basically, to make a cached merkle tree, you get the merkle root of a file every X segments, where X is a power of 2. The cached merkle tree needs to have every element at the same level, and every element needs to be complete.

So, in  the context of Sia, this means that if we're caching at the 4mb level, every single item thrown into a cached merkle tree needs to be fully padded to 4mb.

So, you have a bunch of roots. Then you can throw the roots into the cached tree and you'll get the same merkle root that you would have if you had just taken the merkle root of the whole file all at once.

This means it's easy to change or delete stuff. Just modify or toss the 4mb cached node that you don't want anymore and then recompute the root using the cached nodes you have saved. No need to have the whole original file either, just need to have the cached nodes and then the data that's being changed.

AND, this supports creating storage proof from cached trees too. It's a multi step process though. First, you have to create a storage proof in the 4mb subtree, and then you need to create the storage proof from the cached nodes. You need to hash the whole 4mb structure, and then you need to hash all the cached nodes, but this is probably 100,000x faster (that's not an exaggeration) than hashing the whole file for sufficiently large files.

I didn't spell check this PR, it's 6am and I'm going to bed.

### CODE REVIEW

**Comment by DavidVorick** [comment]
Ok. I'm done now.

File: README.md [hunk 1]
```diff
@@ -46,7 +46,54 @@ func main() {
...
+	// be provided by tree.Prove()
 	merkleRoot, proof, proofIndex, numLeaves := tree.Prove()
+
+	////////////////////////////////////////////////
+	/// Remaining examples deal with cached trees //
+	////////////////////////////////////////////////
+
+	// Example 4: Creating a cached set of Merkle roots and then using them in
+	// a cached tree. The cached tree is height 1, meaning that all elements of
+	// the cached tree will have be merkle roots of data with 2 leaves.
```
Comment (lukechampine) [inline README.md]: typo. also, Merkle is always capitalized

File: README.md [hunk 2]
```diff
@@ -46,7 +46,54 @@ func main() {
...
+	subtree2 := merkletree.New(sha256.New())
+	subtree2.Push([]byte("first leaf, second subtree"))
+	subtree2.Push([]byte("second leaf, second subtree"))
+	// Using the cached tree, build the merkle root of the 4 leaves.
+	cachedTree.Push(subtree1.Root())
+	cachedTree.Push(subtree2.Root())
+	collectiveRoot := cachedTree.Root()
+
+	// Example 5: Modify the data pushed into subtree 2 and create the Merkle
+	root, without needing to rehash the data in any other subtree.
```
Comment (lukechampine) [inline README.md]: missing `//`

File: README.md [hunk 3]
```diff
@@ -46,7 +46,54 @@ func main() {
...
+	revisedSubtree2 := merkletree.New(sha256.New())
+	revisedSubtree2.Push([]byte("first leaf, second subtree"))
+	revisedSubtree2.Push([]byte("second leaf, second subtree, revised"))
+	// Using the cached tree, build the merkle root of the 4 leaves - without
+	// needing to rehash any of the data in subtree1.
+	cachedTree = merkletree.NewCachedTree(sha256.New(), 1)
+	revisedRoot := cachedTree.Root()
```
Comment (lukechampine) [inline README.md]: is this example incomplete? I don't see `revisedSubtree2` being used anywhere.

File: cachedtree.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+// A CachedTree can be used to build Merkle roots and proofs from the cached
+// Merkle roots of smaller blocks of data. Each CachedTree has a height,
+// meaning every element added to the CachedTree is the root of a full Merkle
+// tree containing 2^height leaves.
+type CachedTree struct {
+	cachedNodeHeight uint64
+
+	trueProofIndex uint64
+
+	Tree
```
Comment (lukechampine) [inline cachedtree.go]: why the spacing?

File: merkletree.go [hunk 5]
```diff
@@ -3,378 +3,20 @@
...
+// building the Merkle root for each 4MB sector and remembering the Merkle
+// roots of each sector. Using a cached tree, the Merkle root of the whole file
+// can be computed by passing the cached tree each of the roots of the 4MB
+// sector. Building proofs using these cached roots is also supported. A proof
+// must be build within the target sector using a normal Tree, requiring the
+// whole sector to be hashed. The results of that proof can then be passed into
+// the Prove() function of a cached tree, which will create the full proof
+// without needing to hash the entire file.
+//
+// Cacheing also makes it inexpensive to update the Merkle root of the file
```
Comment (lukechampine) [inline merkletree.go]: typo

File: merkletree.go [hunk 6]
```diff
@@ -3,378 +3,20 @@
...
 // tree of a given root. The tree is implemented according to the specification
 // for Merkle trees provided in RFC 6962.
+//
+// Package merkletree also supports building roots and proofs from cached
+// subroots of the Merkle tree. For example, a large file could be cached by
+// building the Merkle root for each 4MB sector and remembering the Merkle
+// roots of each sector. Using a cached tree, the Merkle root of the whole file
+// can be computed by passing the cached tree each of the roots of the 4MB
+// sector. Building proofs using these cached roots is also supported. A proof
+// must be build within the target sector using a normal Tree, requiring the
```
Comment (lukechampine) [inline merkletree.go]: typo

File: verify.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	// Return false for nonsense input. A switch statement is used so that the
+	// cover tool will reveal if a case is not covered by the test suite. This
+	// would not be possible using a single if statement due to the limitations
+	// of the cover tool.
+	switch {
+	case merkleRoot == nil:
+		return false
+	case proofIndex >= numLeaves:
+		return false
+	}
```
Comment (lukechampine) [inline verify.go]: wouldn't an `if/else if` work just as well? Unless you foresee adding a bunch more conditionals.

File: readers.go [hunk 8]
```diff
@@ -50,7 +50,10 @@ func ReaderRoot(r io.Reader, h hash.Hash, segmentSize int) (root []byte, err err
 // there are not enough bytes remaining in the reader.
 func BuildReaderProof(r io.Reader, h hash.Hash, segmentSize int, index uint64) (root []byte, proofSet [][]byte, numLeaves uint64, err error) {
 	tree := New(h)
-	tree.SetIndex(index)
+	err = tree.SetIndex(index)
+	if err != nil {
+		return
+	}
```
Comment (lukechampine) [inline readers.go]: should probably be a `panic` since this error is unreachable

File: README.md (hunk 3 above)
Comment (DavidVorick) [inline README.md]: ah darn, forgot to push the subtrees into the cached tree.

File: merkletree.go (hunk 6 above)
Comment (DavidVorick) [inline merkletree.go]: I promise I ran spellcheck

File: verify.go (hunk 7 above)
Comment (DavidVorick) [inline verify.go]: this code is from an earlier time, at some point I read that you shouldn't use if/else at all. But yes, that should be good enough I don't forsee this code changing or getting more conditionals at any point.

File: merkletree.go (hunk 6 above)
Comment (lukechampine) [inline merkletree.go]: spellcheck doesn't catch grammatical errors :P

Comment (DavidVorick) [inline merkletree.go]: :set spell=yes --Champine=enabled

**Pull request merged**